CSC Beam Halo filter update

lathomas opened the pull request:
CSC Beam Halo Filter update. 
-Pairs of segments in opposite endcaps are now matched with a dr/dz condition (rather than dr previously) 
-New flat CSC segment matching added (the old one remains the default for now): unvetoing CSC segments associated to tracker muons with pt < 3 gev 
-Optional dt matching of opposite endcaps segments 
-Testing a CSC segment/calorechit  matching. 
-Two new working points added in BeamHaloSummary.

lathomas committed:
CSC Beam Halo filter update

cmsbuild commented:
A new Pull Request was created by @lathomas  for CMSSW_7_4_X.

CSC Beam Halo filter update

It involves the following packages:

DataFormats/METReco
RecoMET/METAlgorithms
RecoMET/METProducers


@cmsbuild, @cvuosalo, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @ahinzmann, @mmarionncern, @jdolen, @nhanvtran, @schoef, @mariadalfonso, @TaiSakuma this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented:
@cmsbuild please test

@lathomas please post a link to a presentation describing the changes

cmsbuild commented:
The tests are being triggered in jenkins.

lathomas committed:
BeamHaloSummary correction

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

slava77 commented:
@lathomas please confirm here in this PR thread if you are done with the changes.
We can't follow a moving target.

cmsbuild commented:
+1
Tested at: aa4ad37af23cf28858b19fb5cd1c3dd34b07a07d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/7898/summary.html

lathomas commented:
Hello,
yes I am done !
Sorry for the trouble.

Laurent
Le 4 sept. 2015 19:15, "Slava Krutelyov" <notifications@github.com> a
écrit :

> @lathomas <https://github.com/lathomas> please confirm here in this PR
> thread if you are done with the changes.
> We can't follow a moving target.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11136#issuecomment-137796206>.
>

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/7898/summary.html

lathomas commented:
Hi,

The changes on the current filter only affect the counting of the flat segments, especially for opposite endcaps.
I didn't mention it in the PR but the dphi condition has been tightened from 0.35 to 0.2.
For opposite endcaps, the dr<10cm matching condition is now replaced by a dr/dz<0.03 condition, which is looser (it corresponds to dr<  36 cm to 60 cm, depending on the chambers).

The current distribution of the two variables entering the current filter (longest chain of flat segments and opposite endcaps segments found) is shown on S15 here (left plots):
https://indico.cern.ch/event/438533/contribution/1/attachments/1139775/1632303/beamhalo.pdf

These are the only changes to the default filter. 

I have also added various methods to the CSCHaloData class. Two new filters were added to BeamHaloSummary but are not used anywhere so far. So all the changes, except the two mentioned above should not break anything and the size of the updated classes should not be dramatically modified. 

Let me know if you need more information. 

Thanks a lot !

Laurent

cvuosalo commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -545,14 +594,67 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
+            dr_rhseg< 25 && dr_rhseg> -25 && //careful: asymmetric cut might not be the most appropriate thing                                  
+	    dtcorr_rhseg>0
+            ) calomatched= true;
+     }
```
If the three "for" loops in lines 602-644 are intended to be the same, they should be made into a function that is called three times, once each for hbhehits, ecalebhits, and ecaleehits. Maybe there is some subtle difference I'm missing. But copy-and-paste programming should be avoided.

cvuosalo commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -561,14 +663,19 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
+	 //	 if(fabs(jZ)<650&& TheEvent.id().run() < 251737)jT-= 25;
+	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= 0.2//max_segment_phi_diff 
+	     //&& TMath::Abs(jR - iR) <= max_segment_r_diff 
+	     && (TMath::Abs(jR - iR) <= max_segment_r_diff || (TMath::Abs(jR - iR)/ TMath::Abs(jZ - iZ)< 0.03 &&  jZ*iZ<0) )
```
Please protect against division by zero if (jZ - iZ) == 0. I don't see any guarantee that such a case would never occur.

cvuosalo commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -579,21 +686,37 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
+             nSegs_alt++;
+             minus_endcap = iGlobalPosition.z() < 0 || jGlobalPosition.z() < 0;
+             plus_endcap = iGlobalPosition.z() > 0 || jGlobalPosition.z() > 0;
+             if( fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )<0.05 && fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )>0.02 && minus_endcap&&plus_endcap ) both_endcaps_loose_dtcut_alt =true;
```
Again, please protect against division by zero here. I'm not sure how likely it is for these quantities to be zero, but it is better to be safe.

slava77 commented on RecoMET/METAlgorithms/src/CSCHaloAlgo.cc:
```diff
@@ -561,14 +663,19 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
+	 float jZ = jGlobalPosition.z() ;
+	 float jT = jSegment->time();
+	 //	 if(fabs(jZ)<650&& TheEvent.id().run() < 251737)jT-= 25;
+	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= 0.2//max_segment_phi_diff 
```
don't use TMath::, unless required: use std:: functions instead
- replace this acos(cos nonsense with ``` std::abs(reco::deltaPhi ( jPhi - iPhi )); ```

lathomas committed:
CSCHaloAlgo update

lathomas commented:
Hello, 
I have implemented your comments. Some of the things reported (like the Acos (cos)) were actually present since a long time but I modified this. 
I protected against zero division. 
I have also written two functions for ECAL and HCAL matching (since the collections to pass are different, there is maybe a better way to do things though...).  
I checked on a beam halo enriched samples that the changes are purely cosmetics (no difference in the filter decision between the two versions). 

I will wait that this PR is approved before updating the 75X one... 
Laurent

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: c6f4cd1cf721b613d11821985b1838f5a53c1da1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8007/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8007/summary.html

cvuosalo commented:
+1

For #11136 c6f4cd1cf721b613d11821985b1838f5a53c1da1

Small improvements to the CSC beam halo filter.

The code changes are satisfactory, and Jenkins [tests](https://github.com/cms-sw/cmssw/pull/11136) against baseline CMSSW_7_4_X_2015-09-08-2300  show no significant differences. Tests of wokflows 8.0_BeamHalo and 25202.0_TTbar_13 against baseline CMSSW_7_4_11_patch1also show no significant differences. CPU times tests for 25202 show no significant increase in time, with a small addition of 12-17 ms/event for CSC halo processing.

One measure of processing time:
```
	CSCHaloData 	 0.296012 ms/ev -> 12.0602 ms/ev
 Total times: 6478.48 ms/ev -> 6501.05 ms/ev 	 delta: 22.5697
```
Other measures of processing time and memory:
```
Excluding the first event:
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.949071      +0.02%         0.22 ms/ev ->        17.12 ms/ev CSCHaloData
Job total:  101.357 s/ev ==> 100.159 s/ev

Baseline
Max VSIZ 3606.48 on evt 34 ; max RSS 3044.13 on evt 45
Time av 102.222 s/evt   max 138.68 s on evt 36
M1 Time av 102.24 s/evt   max 138.68 s on evt 36

PR 
Max VSIZ 3606.25 on evt 38 ; max RSS 3018.09 on evt 45
Time av 102.465 s/evt   max 201.66 s on evt 1
M1 Time av 101.027 s/evt   max 137.612 s on evt 18
```

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs once checked with relvals in the development release cycle of CMSSW (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

lathomas committed:
Update CSC Halo Filter

lathomas commented:
@cvuosalo  @slava77  @davidlange6
I have updated this PR to synchronize it with the recent 76X PR update (now merged): 
https://github.com/cms-sw/cmssw/pull/11528

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 019f21d9859754760ae00a969589162a83140364
 When I ran the RelVals I found an error in the following worklfows: 
 4.22 step2
<pre>runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC/step2_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC.log

</pre>
4.53 step2
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT/step2_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT.log

</pre>
5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>
8.0 step1
<pre>runTheMatrix-results/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/step1_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS.log

</pre>
9.0 step1
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step1_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>
25.0 step1
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaEle+HARVEST+ALCATT/step1_TTbar+TTbar+DIGI+RECOAlCaEle+HARVEST+ALCATT.log

</pre>
135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>
1306.0 step1
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step1_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
1330.0 step1
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step1_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>
101.0 step1
<pre>runTheMatrix-results/101.0_SingleElectronE120EHCAL+SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL+SingleElectronE120EHCAL.log

</pre>
25202.0 step1
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step1_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>
50202.0 step1
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step1_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>
1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>
1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>
1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4/step2_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8564/summary.html

lathomas committed:
Removing  Halo Strip Filter

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f59487aa12fbacbbbbe6fe2b8f67f0fc3809170a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8569/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8569/summary.html

cvuosalo commented:
+1

For #11136 f59487aa12fbacbbbbe6fe2b8f67f0fc3809170a

Second approval, after addition of analysis code from #11528. New Jenkins tests still show no significant differences, and a repeat test of workflow 25202.0 against baseline CMSSW_7_4_11_patch1 also shows no significant differences and similar timing results as before (see previous approval comment).

lathomas committed:
Tuning the various parameters entering the filter to decrease the fake rate, only consider out of time hits in the calo-csc matching)

lathomas commented:
Copy of the following PR for 75X: 
https://github.com/cms-sw/cmssw/pull/11737

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
JENKINS_TEST_URL

cmsbuild commented:
+1
Tested at: bf4d5c138a70d7caed944777c60bccbc5a02613d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8772/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11136/8772/summary.html

cvuosalo commented:
+1

For #11136 bf4d5c138a70d7caed944777c60bccbc5a02613d

Third approval of this PR for new, small adjustments to the CSC beam halo filter that come from 76X #11737, which has already been merged. #11137 is the matching PR to this one for 75X, and it has already been approved by Reco.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_7_4_X_2015-10-12-2300 show no significant differences. An additional test of workflow 25202.0_TTbar_13 was done for #11737, and it showed no significant differences.

lathomas committed:
Typo in  CSCTightHalo2015Filter

lathomas committed:
Reload calogeometry for every event: geo object seems to change its memory address between runs, leading to segmentation fault from time to time, the change should hopefully fix this

slava77 commented:
@davidlange6 
I think that this PR can be closed
Even if we do something with 2015, it will be 76X.

cmsbuild commented:
Pull request #11136 was updated. @cmsbuild, @cvuosalo, @slava77, @monttj, @davidlange6 can you please check and sign again.

slava77 commented:
-1
to be closed
